CMakeLists.txt

```diff
@@ -0,0 +1,250 @@
+#   library combination. See OTIO.
+
+# Build options
+
+# Surface `BUILD_SHARED_LIBS` as a non-hidden CMake option in CMake
+# GUIs. This controls whether static or shared libs are built.
+# TODO(DF): We may end up needing to support building of both shared
+#   and static libs simultaneously. Also, using this means we inherit
+#   the option if openassetio is used as a CMake subproject.
+option(BUILD_SHARED_LIBS "Set to OFF to build static libraries" ON)
+
+# Default treating compiler warnings as errors to OFF, since
+# consumers of this project may use unpredictable toolchains.
+# For dev/CI we should remember to switch this ON, though!
+option(OPENASSETIO_WARNINGS_AS_ERRORS "Treat compiler warnings as errors" OFF)
+
+# Enable IPO by default. For such a small project, the increased build
+# resources seem worth the potential additional compiler optimisations.
+option(OPENASSETIO_ENABLE_IPO
+    "Enable Interprocedural Optimization, aka Link Time Optimization (LTO)" ON)
+
+# Enable position independent code by default. This is redundant for
+# shared libraries (CMake sets it automatically). Enabling this for
+# static libraries allows symbols to be exported by downstream
+# shared libraries.
+option(OPENASSETIO_ENABLE_POSITION_INDEPENDENT_CODE
+    "Enable position independent code for static library builds" ON)
+
+# Disable C++11 ABI by default, as per current VFX reference
+# platform(s), but allow optionally enabling.
+# For reference, note that the default varies by platform, e.g.
+# * CentOS 7.9.2009's GCC 6.3, `_GLIBCXX_USE_CXX11_ABI` is
+#   always `0` and cannot be overridden.
+# * Ubuntu 18.04's GCC 6.5, `_GLIBCXX_USE_CXX11_ABI` defaults
+#   to `1`, but can be overridden.
+if (CMAKE_CXX_COMPILER_ID STREQUAL "GNU" AND
+    CMAKE_CXX_COMPILER_VERSION VERSION_GREATER_EQUAL 5.0)
+    option(OPENASSETIO_GLIBCXX_USE_CXX11_ABI "For gcc, use the new C++11 library ABI" OFF)
+endif ()
+
+# Whether to use RPATH or RUNPATH runtime search path behavior.
+if (IS_GCC_OR_CLANG)
+    option(OPENASSETIO_ENABLE_NEW_DTAGS "If enabled, RUNPATH will be set, overriding RPATH" OFF)
+endif ()
+
+# Testing instrumentation.
+if (IS_GCC_OR_CLANG)
+    # Coverage reporting (for tests).
+    option(OPENASSETIO_ENABLE_COVERAGE "Enable coverage reporting for gcc/clang" OFF)
+
+    # Sanitizers (primarily for tests)
+    option(OPENASSETIO_ENABLE_SANITIZER_ADDRESS "Enable address sanitizer" OFF)
+    option(OPENASSETIO_ENABLE_SANITIZER_LEAK "Enable leak sanitizer" OFF)
+    option(OPENASSETIO_ENABLE_SANITIZER_UNDEFINED_BEHAVIOR
+        "Enable undefined behavior sanitizer" OFF)
+    option(OPENASSETIO_ENABLE_SANITIZER_THREAD "Enable thread sanitizer" OFF)
+    option(OPENASSETIO_ENABLE_SANITIZER_MEMORY "Enable memory sanitizer" OFF)
+endif ()
+
+
+#-----------------------------------------------------------------------
+# Enable ctest testing
+
+# To correctly generate the DartConfiguration.tcl, include the CTest
```

Comment:
Missed this on the first pass.

What's the `DartConfiguration.tcl`?

Reply:
This came from [OCIO's CMakeLists.txt](https://github.com/AcademySoftwareFoundation/OpenColorIO/blob/1d8a28c2e2f842de053cef25887b7d5bc89dc438/CMakeLists.txt#L43). That file is generated when you include `CTest`. Its presumably consumed by the `ctest` tool, and apparently required for some of its built-in utilities (i.e.[memcheck](https://cmake.org/cmake/help/latest/manual/ctest.1.html#ctest-memcheck-step)).  

I'm not sure how we'll configure testing infrastructure. I left in that comment since `ctest` may be useful when we come to look at organising different types of test.

If interested, the contents of the generated `DartConfiguration.tcl` are
```
# This file is configured by CMake automatically as DartConfiguration.tcl
# If you choose not to use CMake, this file may be hand configured, by
# filling in the required variables.


# Configuration directories and files
SourceDirectory: /home/dave/workspace/cloud/assetapi/OpenAssetIO
BuildDirectory: /home/dave/workspace/cloud/assetapi/OpenAssetIO/cmake-build-debug

# Where to place the cost data store
CostDataFile: 

# Site is something like machine.domain, i.e. pragmatic.crd
Site: FWTP5313

# Build name is osname-revision-compiler, i.e. Linux-2.4.2-2smp-c++
BuildName: Linux-g++-6

# Subprojects
LabelsForSubprojects: 

# Submission information
SubmitURL: http://

# Dashboard start time
NightlyStartTime: 00:00:00 EDT

# Commands for the build/test/submit cycle
ConfigureCommand: "/usr/local/lib/python3.8/dist-packages/cmake/data/bin/cmake" "/home/dave/workspace/cloud/assetapi/OpenAssetIO"
MakeCommand: /usr/local/lib/python3.8/dist-packages/cmake/data/bin/cmake --build . --config "${CTEST_CONFIGURATION_TYPE}"
DefaultCTestConfigurationType: Release

# version control
UpdateVersionOnly: 

# CVS options
# Default is "-d -P -A"
CVSCommand: CVSCOMMAND-NOTFOUND
CVSUpdateOptions: -d -A -P

# Subversion options
SVNCommand: /usr/bin/svn
SVNOptions: 
SVNUpdateOptions: 

# Git options
GITCommand: /usr/bin/git
GITInitSubmodules: 
GITUpdateOptions: 
GITUpdateCustom: 

# Perforce options
P4Command: /usr/bin/p4
P4Client: 
P4Options: 
P4UpdateOptions: 
P4UpdateCustom: 

# Generic update command
UpdateCommand: /usr/bin/git
UpdateOptions: 
UpdateType: git

# Compiler info
Compiler: /usr/bin/g++-6
CompilerVersion: 6.5.0

# Dynamic analysis (MemCheck)
PurifyCommand: 
ValgrindCommand: 
ValgrindCommandOptions: 
DrMemoryCommand: 
DrMemoryCommandOptions: 
MemoryCheckType: 
MemoryCheckSanitizerOptions: 
MemoryCheckCommand: MEMORYCHECK_COMMAND-NOTFOUND
MemoryCheckCommandOptions: 
MemoryCheckSuppressionFile: 

# Coverage
CoverageCommand: /usr/bin/gcov
CoverageExtraFlags: -l

# Cluster commands
SlurmBatchCommand: SLURM_SBATCH_COMMAND-NOTFOUND
SlurmRunCommand: SLURM_SRUN_COMMAND-NOTFOUND

# Testing options
# TimeOut is the amount of time in seconds to wait for processes
# to complete during testing.  After TimeOut seconds, the
# process will be summarily terminated.
# Currently set to 25 minutes
TimeOut: 1500

# During parallel testing CTest will not start a new test if doing
# so would cause the system load to exceed this value.
TestLoad: 

UseLaunchers: 
CurlOptions: 
# warning, if you add new options here that have to do with submit,
# you have to update cmCTestSubmitCommand.cxx

# For CTest submissions that timeout, these options
# specify behavior for retrying the submission
CTestSubmitRetryDelay: 5
CTestSubmitRetryCount: 3
```

Reply:
Interesting. We use CTest a lot in the Conan recipes, but never used this.
I now see it in https://gitlab.kitware.com/cmake/community/-/wikis/doc/ctest/Generating-Testing-Files

Since I've not had any experience of CTest using this, if you think it'll be useful to you, ignore me.

(Personally, I'd only add it when I had some tests to run.)

Reply:
Removed reference to the CTest module, for now, in c0dd3eeb.  

We'll definitely need testing infrastructure, but the CTest CDash client is probably overkill for our little project on github anyway - it [looks like](https://www.kitware.com/github-checks-and-statuses-from-cdash/) you need a separate server for your builds, which then notifies github.